plutus-metatheory/src/Cost/Size.lagda.md

```diff
@@ -0,0 +1,114 @@
+# Size of Constants
+
+```
+module Cost.Size where 
+
+```
+
+# Imports
+
+```
+open import Data.Nat using (ℕ;zero;suc;_+_;_*_;_∸_)
+open import Data.Nat.DivMod using (_/_)
+open import Data.Integer using (ℤ;∣_∣)
+open import Data.String using (length)
+
+open import Utils using (_×_;_,_;_∷_;[];DATA;List;ByteString)
+open DATA
+
+open import Builtin using (lengthBS)
+open import Builtin.Signature using (_⊢♯)
+open _⊢♯
+
+open import Builtin.Constant.AtomicType using (AtomicTyCon)
+open AtomicTyCon
+open import Cost.Base 
+open import Untyped.CEK using (Value;V-con)
+open import RawU using (TmCon;tmCon) 
+
+open import Data.Vec using (Vec;[];_∷_) 
+                     renaming (lookup to lookupVec)
+
+```
+
+## Memory usage of type constants
+
+For each type constant we calculate its size, as a measure of memory usage.
+
+First we bring some functions from Haskell world.
+
+```
+postulate ℕtoWords : ℤ → CostingNat 
+postulate g1ElementCost : CostingNat
+postulate g2ElementCost : CostingNat
+postulate mlResultElementCost : CostingNat 
+
+{-# FOREIGN GHC {-# LANGUAGE MagicHash #-} #-}
+{-# FOREIGN GHC import GHC.Exts (Int (I#)) #-}
+{-# FOREIGN GHC import GHC.Integer  #-}
+{-# FOREIGN GHC import GHC.Integer.Logarithms  #-}
+{-# FOREIGN GHC import GHC.Prim #-}
+{-# FOREIGN GHC import PlutusCore.Crypto.BLS12_381.G1 as BLS12_381.G1 #-}
+{-# FOREIGN GHC import PlutusCore.Crypto.BLS12_381.G2 as BLS12_381.G2 #-}
+{-# FOREIGN GHC import PlutusCore.Crypto.BLS12_381.Pairing as BLS12_381.Pairing #-}
+
+{-# COMPILE GHC ℕtoWords = \i -> fromIntegral $ I# (integerLog2# (abs i) `quotInt#` integerToInt 64) + 1 #-}
```

Comment:
There may be a tiny danger of incompatibility for the size of 0 here because of [this](https://github.com/IntersectMBO/plutus/blob/4da3e0ad032209d6cb6e9aa4c0c9e2f2325d4763/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs#L189), which we only discovered a few days ago.  I don't think it's a big deal though.

Reply:
Fixed. Now returns 1 for i == 0.